sympy/solvers/solveset.py

```diff
@@ -532,6 +532,80 @@ def _solve_abs(f, symbol, domain):
         return ConditionSet(symbol, Eq(f, 0), domain)
 
 
+def solve_decomposition(f, symbol, domain):
+    """
+    Function to solve equations via the principle of "Decomposition
+    and Rewriting".
+
+    Examples
+    ========
+    >>> from sympy import exp, sin, Symbol, pprint, S
+    >>> from sympy.solvers.solveset import solve_decomposition as sd
+    >>> x = Symbol('x')
+    >>> f1 = exp(2*x) - 3*exp(x) + 2
+    >>> sd(f1, x, S.Reals)
+    {0, log(2)}
+    >>> pprint(sd(f1, x, S.Complexes), use_unicode=False)
+    {2*n*I*pi | n in Integers()} U {2*n*I*pi + Mod(log(2), 2*I*pi) | n in Integers
+    ()}
+    >>> f2 = sin(x)**2 + 2*sin(x) + 1
+    >>> pprint(sd(f2, x, S.Reals), use_unicode=False)
+              3*pi
+    {2*n*pi + ---- | n in Integers()}
+               2
+    >>> f3 = sin(x + 2)
+    >>> pprint(sd(f3, x, S.Reals), use_unicode=False)
+    {2*n*pi + 2 | n in Integers()} U {2*n*pi + 2 + pi | n in Integers()}
+
+    """
+    from sympy.solvers.decompogen import decompogen
+    # decompose the given function
+    g_s = decompogen(f, symbol)
+    # `y_s` represents the set of values for which the function `g` is to be
+    # solved.
+    # `solutions` represent the solutions of the equations `g = y_s` or
+    # `g = 0` depending on the type of `y_s`.
+    # As we are interested in solving the equation: f = 0
+    y_s = FiniteSet(0)
+    for g in g_s:
+        # initialise `results` as an empty set at the beginning of iteration
+        result = S.EmptySet
+        if isinstance(y_s, FiniteSet):
+            for y in y_s:
+                solutions = solveset(Eq(g, y), symbol, domain)
+                if not isinstance(solutions, ConditionSet):
+                    result += solutions
+        else:
+            solutions = solveset(g, symbol, domain)
+            # if isinstance(solutions, ConditionSet) or solutions is S.EmptySet:
```

Comment:
This is just an initial commit.
Also, the comments here are aimed to be guidelines for better understanding of the method.
I will make sure of this when the PR will be ready for review.
Thanks. :+1: